jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUTreeIsomorphismInspector.java

```diff
@@ -0,0 +1,349 @@
+/*
+ * (C) Copyright 2018-2018, by Alexandru Valeanu and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.isomorphism;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphMapping;
+import org.jgrapht.Graphs;
+import org.jgrapht.alg.shortestpath.TreeMeasurer;
+import org.jgrapht.alg.util.Pair;
+import org.jgrapht.traverse.BreadthFirstIterator;
+import org.jgrapht.util.RadixSort;
+
+import java.util.*;
+
+/**
+ * This is an implementation of the AHU algorithm for detecting an (unweighted) isomorphism between two trees.
+ * Please see <a href="http://mathworld.wolfram.com/GraphIsomorphism.html">mathworld.wolfram.com</a> for a complete
+ * definition of the isomorphism problem for general graphs.
+ *
+ * <p>
+ *     The original algorithm was first presented in "Alfred V. Aho and John E. Hopcroft. 1974.
+ *     The Design and Analysis of Computer Algorithms (1st ed.). Addison-Wesley Longman Publishing Co., Inc., Boston, MA, USA."
+ * </p>
+ *
+ * <p>
+ *     This implementation runs in linear time (in the number of vertices of the input trees)
+ *     while using a linear amount of memory.
+ * </p>
+ *
+ * <p>
+ *      For an implementation that supports rooted forests see {@link AHUForestIsomorphismInspector}.
+ * </p>
+ *
+ * <p>
+ *     Note: This inspector only returns a single mapping (chosen arbitrarily) rather than all possible mappings.
+ * </p>
+ *
+ * @param <V> the type of the vertices
+ * @param <E> the type of the edges
+ *
+ * @author Alexandru Valeanu
+ */
+public class AHUTreeIsomorphismInspector<V, E> implements IsomorphismInspector<V, E> {
+    private final Graph<V, E> tree1;
+    private final Graph<V, E> tree2;
+
+    private V root1;
+    private V root2;
+
+    private Map<V, V> forwardMapping;
+    private Map<V, V> backwardMapping;
+
+    /**
+     * Construct a new AHU unrooted tree isomorphism inspector.
+     *
+     * Note: The constructor does NOT check if the input trees are valid.
+     *
+     * @param tree1 the first tree
+     * @param tree2 the second tree
+     * @throws NullPointerException if {@code tree1} is {@code null}
+     * @throws NullPointerException if {@code tree2} is {@code null}
+     */
+    public AHUTreeIsomorphismInspector(Graph<V, E> tree1, Graph<V, E> tree2){
```

Comment:
This is very strange. This constructor allows null root values, but the other does not.

Use `this(tree1, null, tree2, null);`

Reply:
It is not strange. One constructor (the one that doesn't take root nodes) is for unrooted isomorphism while the other is for the rooted case. It makes sense for the one for the rooted case to check for nullness and reject the input.

Reply:
Constructors in classes should not have different behavior. You should have one constructor which accepts all parameters and perhaps some others for special cases. The special cases should call the more generic one. This way you avoid bugs.

Reply:
In this case, should I have two classes, one for solving the rooted tree iso and one for the unrooted tree iso problem? I still believe that if we properly document the two cases (and constructors) one class can be used for both. @jkinable @jsichi any opinions on this?

Reply:
After looking the implementation in more detail, I feel that two classes would be better.

You won't even have to use inheritance, since your code is a reduction from the unrooted to the rooted case (with the center(s) as roots).

Reply:
I agree that two classes would be best, especially since for the unrooted case, we should require the input to be undirected, whereas for the rooted case, either directed or undirected is fine.